File: Form1.vb
Imports System
Imports System.Windows.Forms

Namespace AdeTeK_Plottertest
    Public Partial Class Form1
        Inherits Form

        Public Shared StepsProMM As Double = 40.0 'Bei 1600 Steps/Umdrehung

        'BEGIN Für die Motorfahrt
        Public Shared sollPosition_X As Integer
        Public Shared sollPosition_Y As Integer
        Public Shared istPosition_X As Integer = 0
        Public Shared istPosition_Y As Integer = 0
        Public Shared stepsToGo_X As Integer = 0
        Public Shared stepsToGo_Y As Integer = 0
        Public Shared X_Steps, Y_Steps, SummeSteps_XY As Integer
        Public Shared sleepPause As Integer
        'END Für die Motorfahrt

        '###########################################################
        Public Delegate Sub SetTextCallback(ByVal text As String)
        '###########################################################
        Public Sub New()
            Me.InitializeComponent()
        End Sub
        '###########################################################
        Private Sub ReceivedText(ByVal text As String)
            If Me.tbReceived.InvokeRequired Then
                Dim x = New SetTextCallback(AddressOf ReceivedText)
                Invoke(x, New Object() {text})
            Else
                Me.tbReceived.Text += text
            End If
        End Sub
        '###########################################################
        Private Sub serialPort1_DataReceived(ByVal sender As Object, ByVal e As Ports.SerialDataReceivedEventArgs)
            Me.ReceivedText(Me.serialPort1.ReadExisting())
        End Sub
        '###########################################################
        Private Sub Sleep(ByVal sleeptime As Integer)
            Dim stopw = New Stopwatch()
            stopw.Start()

            While stopw.ElapsedMilliseconds < sleeptime
                Call Application.DoEvents()
            End While

            stopw.Stop()
            stopw.Reset()
        End Sub

        Private Sub MsgBox(ByVal v As String)
            MsgBox("COM Port nicht gefunden...")
        End Sub
        '###########################################################
        Private Sub btnConnect_Click(ByVal sender As Object, ByVal e As EventArgs)
            Try
                Me.serialPort1.Open() '(Exception ex)
            Catch
                MsgBox("COM Port nicht gefunden...")
            End Try
            'timer1.Start();
            Me.panel1.Enabled = True
            Me.btnConnect.Enabled = False
            Me.btnDisconnect.Enabled = True
        End Sub
        '-----------------------------------------------------------
        Private Sub timer1_Tick(ByVal sender As Object, ByVal e As EventArgs)
            'try
            '{
            '    string Text = serialPort1.ReadExisting();
            '    tbReceived.Text += Text;
            '    tbxTest1.Text += Text;
            '}
            'catch //(Exception ex)
            '{
            '}           
        End Sub
        '-----------------------------------------------------------
        Private Sub btnDisconnect_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.serialPort1.Close()
            Me.panel1.Enabled = False
            Me.btnConnect.Enabled = True
            Me.btnDisconnect.Enabled = False
        End Sub
        '-----------------------------------------------------------
        Private Sub BtnSend_Click(ByVal sender As Object, ByVal e As EventArgs)
            Try
                Me.serialPort1.Write(Me.tbTxtTransmit.Text) ' WriteLine und nicht nur Write!!!!!
                '(Exception ex)
            Catch
                MsgBox("COM Port nicht verbunden")
            End Try
        End Sub
        '-----------------------------------------------------------
        Private Sub btnExit_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.serialPort1.Close()
            Close()
        End Sub
        '-----------------------------------------------------------
        Private Sub btnReset_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.serialPort1.Write("2,0,0,0,0")      'Reset    case 2:
        End Sub
        'case 5: Arduino
        Private Sub btnServoUp_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.serialPort1.Write("5,0,0,0,0")    'ServoUp    case 5:
        End Sub

        Private Sub btnServoDown_Click(ByVal sender As Object, ByVal e As EventArgs)
            'case 6: Arduino
            Me.serialPort1.Write("6,0,0,0,0")   'ServoDown   case 6:
        End Sub
        '-----------------------------------------------------------
        'case 7: Arduino
        Private Sub btnMagnetON_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.serialPort1.Write("7,0,0,0,0")     'MagnetON case 7:
        End Sub
        '-----------------------------------------------------------
        'case 8: Arduino
        Private Sub btnMagnetOFF_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.serialPort1.Write("8,0,0,0,0")    'MagnetOFF case 8:
        End Sub
        '-----------------------------------------------------------
        Private Sub btnPickUpFigure_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.serialPort1.Write("6,0,0,0,0") 'ServoDown case 6:
            Sleep(2000)
            Me.serialPort1.Write("7,0,0,0,0") 'MagnetON  case 7:
            Sleep(2000)
            Me.serialPort1.Write("5,0,0,0,0") 'ServoUp   case 5:
        End Sub
        '-----------------------------------------------------------
        Private Sub btnPutDownFigure_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.serialPort1.Write("6,0,0,0,0") 'ServoDown case 6:
            Sleep(2000)
            Me.serialPort1.Write("8,0,0,0,0") 'MagnetOFF case 8:
            Sleep(2000)
            Me.serialPort1.Write("5,0,0,0,0") 'ServoUp   case 5:
        End Sub
        '-----------------------------------------------------------
        Private Sub btnTbReceivedClear_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.tbReceived.Clear()
            Me.tbxTest1.Clear()
            Me.tbxTest2.Clear()
            Me.tbxTest3.Clear()
            Me.tbxTest4.Clear()
        End Sub
        '-----------------------------------------------------------
        Private Sub btn00_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(0, 0)
        End Sub

        Private Sub btn1600_1600_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(1600, 1600)
        End Sub

        Private Sub btn6000_1000_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(6000, 1000)
        End Sub

        Private Sub btn3200_1600_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(3200, 1600)
        End Sub

        Private Sub btn2000_1000_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(2000, 1000)
        End Sub

        Private Sub btn4000_800_Click(ByVal sender As Object, ByVal e As EventArgs) 'geändert in 2000/4000
            FahrtXY(2000, 4000)
        End Sub

        Private Sub btn9900_700_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(9900, 700)
        End Sub

        Private Sub btnTextboxSend_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(5000, 3000)
        End Sub

        Private Sub btnLabelSend_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(2000, 5500)
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
        End Sub
        '-----------------------------------------------------------
        Private Sub btnTextBoxTest_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.tbReceived.Clear()
            Me.tbxTest1.Clear()
            Me.tbxTest2.Clear()
            Me.tbxTest3.Clear()
            Me.tbxTest4.Clear()
            Me.serialPort1.Write("14,0,0,0,0")
            Sleep(1500)
            Dim Zeile = Me.tbReceived.Lines

            Try
                Me.tbxTest1.Text = Zeile(0)
                Me.tbxTest2.Text = Zeile(1)
                Me.tbxTest3.Text = Zeile(2)
                Me.tbxTest4.Text = Zeile(3) '(Exception ex)
            Catch
                MsgBox("Es sind noch keine Werte vorhanden")
            End Try
        End Sub
        '-----------------------------------------------------------
        Private Sub btnTestVar_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(4650, 7280)
        End Sub
        '-----------------------------------------------------------
        Private Sub txbRunden1_MouseDoubleClick(ByVal sender As Object, ByVal e As MouseEventArgs)
            Me.serialPort1.Write("15," & Me.txbRunden1.Text & ",0,0,0")
        End Sub
        '-----------------------------------------------------------
        Private Sub btnStaticTime_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim Pausenzeit = 5000
            Me.tbxTest3.Text = "Pause: " & Pausenzeit.ToString()
            Me.serialPort1.Write("1,9000,8300,0,0")
            Sleep(Pausenzeit)
            Me.serialPort1.Write("1,0,0,0,0")
            Sleep(Pausenzeit)
            Me.serialPort1.Write("1,5000,7900,0,0")
            Sleep(Pausenzeit)
            Me.serialPort1.Write("1,0,0,0,0")
        End Sub
        '-----------------------------------------------------------
        Private Sub txbRunden1_KeyPress(ByVal sender As Object, ByVal e As KeyPressEventArgs)
            If e.KeyChar = 13 Then
                Me.serialPort1.Write("15," & Me.txbRunden1.Text & ",0,0,0")
                e.Handled = True 'DingDong ausschalten...
            End If
        End Sub

        Public Shared Function map(ByVal value As Integer, ByVal fromLow As Integer, ByVal fromHigh As Integer, ByVal toLow As Integer, ByVal toHigh As Integer) As Integer
            Return (value - fromLow) * (toHigh - toLow) / (fromHigh - fromLow) + toLow
        End Function 'END private static int map()------------------------------
        '-----------------------------------------------------------
        Private Sub btnWhileSchleife_Click(ByVal sender As Object, ByVal e As EventArgs)
            FahrtXY(9000, 8300)
            Sleep(sleepPause)
            FahrtXY(0, 0)
            Sleep(sleepPause)
            FahrtXY(5000, 7900)
            Sleep(sleepPause)
            FahrtXY(0, 0)
            Sleep(sleepPause)
            FahrtXY(1200, 300)
            Sleep(sleepPause)
            FahrtXY(0, 0)
            Sleep(sleepPause)
        End Sub
        '-----------------------------------------------------------
        Public Sub FahrtXY(ByVal Steps_X As Integer, ByVal Steps_Y As Integer)
            sollPosition_X = Steps_X
            sollPosition_Y = Steps_Y
            stepsToGo_X = Math.Abs(sollPosition_X - istPosition_X)
            stepsToGo_Y = Math.Abs(sollPosition_Y - istPosition_Y)
            SummeSteps_XY = stepsToGo_X + stepsToGo_Y
            Me.tbxSumSteps.Text = SummeSteps_XY.ToString()
            '#######################################################
            'this.serialPort1.Write("1,Steps_X, Steps_Y,0,0"); //--> So funktioniert es nicht...
            Me.serialPort1.Write("1," & Steps_X & "," & Steps_Y & ",0,0") 'zur seriellen Schnittstelle
            '#######################################################
            istPosition_X = sollPosition_X
            istPosition_Y = sollPosition_Y
            sleepPause = map(SummeSteps_XY, 500, 10000, 1000, 4500)
            'tbxSleepPause.Text = sleepPause.ToString();
            Me.tbxTest4.Text = "Pause: " & sleepPause.ToString()
        End Sub
        '-----------------------------------------------------------

    End Class 'END public partial class Form1 : Form
End Namespace 'END namespace AdeTeK_Plottertest

File: Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace AdeTeK_Plottertest
{
    public partial class Form1 : Form
    {
        public static Double StepsProMM = 40.0; //Bei 1600 Steps/Umdrehung

        //BEGIN Für die Motorfahrt
        public static int sollPosition_X;
        public static int sollPosition_Y;
        public static int istPosition_X = 0;
        public static int istPosition_Y = 0;
        public static int stepsToGo_X = 0;
        public static int stepsToGo_Y = 0;
        public static int X_Steps, Y_Steps, SummeSteps_XY;
        public static int sleepPause;
        //END Für die Motorfahrt

        //###########################################################
        public delegate void SetTextCallback(string text);
        //###########################################################
        public Form1()
        {
            InitializeComponent();
        }
        //###########################################################
        private void ReceivedText(string text)
        {
            if (this.tbReceived.InvokeRequired)
            {
                var x = new SetTextCallback(ReceivedText);
                this.Invoke(x, new object[] { text });
            }
            else
            {
                this.tbReceived.Text += text;
            }
        }
        //###########################################################
        private void serialPort1_DataReceived(object sender, System.IO.Ports.SerialDataReceivedEventArgs e)
        {
            this.ReceivedText(this.serialPort1.ReadExisting());
        }
        //###########################################################
        private void Sleep(int sleeptime)
        {
            var stopw = new System.Diagnostics.Stopwatch();
            stopw.Start();
            while (stopw.ElapsedMilliseconds < sleeptime)
                Application.DoEvents();
            stopw.Stop();
            stopw.Reset();
        }
        private void MsgBox(string v)
        {
            MsgBox("COM Port nicht gefunden...");
        }
        //###########################################################
        private void btnConnect_Click(object sender, EventArgs e)
        {
            try
            {
                serialPort1.Open();
            }
            catch //(Exception ex)
            {
                MsgBox("COM Port nicht gefunden...");
            }
            //timer1.Start();
            panel1.Enabled = true;
            btnConnect.Enabled = false;
            btnDisconnect.Enabled = true;
            serialPort1.Write("6,0,0,0,0"); //ServoDown case 6:
            Sleep(2000);
            serialPort1.Write("5,0,0,0,0"); //ServoUp   case 5:
            txbServoUpdatelAngleMin.Text = "10";
            txbServoUpdatelAngleMax.Text = "170";
        }          
        //-----------------------------------------------------------
        private void timer1_Tick(object sender, EventArgs e)
        {
            //try
            //{
            //    string Text = serialPort1.ReadExisting();
            //    tbReceived.Text += Text;
            //    tbxTest1.Text += Text;
            //}
            //catch //(Exception ex)
            //{
            //}           
        }
        //-----------------------------------------------------------
        private void btnDisconnect_Click(object sender, EventArgs e)
        {
            serialPort1.Close();
            panel1.Enabled = false;
            btnConnect.Enabled = true;
            btnDisconnect.Enabled = false;
        }
        //-----------------------------------------------------------
        private void BtnSend_Click(object sender, EventArgs e)
        {
            try
            {
                serialPort1.Write(tbTxtTransmit.Text); // WriteLine und nicht nur Write!!!!!
            }
            catch //(Exception ex)
            {
                MsgBox("COM Port nicht verbunden");
            }
        }
        //-----------------------------------------------------------
        private void btnExit_Click(object sender, EventArgs e)
        {
            serialPort1.Close();
            this.Close();
        }
        //-----------------------------------------------------------
        private void btnReset_Click(object sender, EventArgs e)
        {
            serialPort1.Write("2,0,0,0,0");      //Reset    case 2:
        }

        private void btnServoUp_Click(object sender, EventArgs e)
        //case 5: Arduino
        {
            serialPort1.Write("5,0,0,0,0");    //ServoUp    case 5:
        }

        private void btnServoDown_Click(object sender, EventArgs e)
        {
            //case 6: Arduino
            serialPort1.Write("6,0,0,0,0");   //ServoDown   case 6:

        }
        //-----------------------------------------------------------
        private void btnMagnetON_Click(object sender, EventArgs e)
        //case 7: Arduino
        {
            serialPort1.Write("7,0,0,0,0");     //MagnetON case 7:
        }
        //-----------------------------------------------------------
        private void btnMagnetOFF_Click(object sender, EventArgs e)
        //case 8: Arduino
        {
            serialPort1.Write("8,0,0,0,0");    //MagnetOFF case 8:
        }
        //-----------------------------------------------------------
        private void btnPickUpFigure_Click(object sender, EventArgs e)
        {
            serialPort1.Write("6,0,0,0,0"); //ServoDown case 6:
            Sleep(2000);
            serialPort1.Write("7,0,0,0,0"); //MagnetON  case 7:
            Sleep(2000);
            serialPort1.Write("5,0,0,0,0"); //ServoUp   case 5:
        }
        //-----------------------------------------------------------
        private void btnPutDownFigure_Click(object sender, EventArgs e)
        {
            serialPort1.Write("6,0,0,0,0"); //ServoDown case 6:
            Sleep(2000);
            serialPort1.Write("8,0,0,0,0"); //MagnetOFF case 8:
            Sleep(2000);
            serialPort1.Write("5,0,0,0,0"); //ServoUp   case 5:
        }
        //-----------------------------------------------------------
        private void btnTbReceivedClear_Click(object sender, EventArgs e)
        {
            tbReceived.Clear();
            tbxTest1.Clear();
            tbxTest2.Clear();
            tbxTest3.Clear();
            tbxTest4.Clear();
        }
        //-----------------------------------------------------------
        private void btn00_Click(object sender, EventArgs e)
        {
            FahrtXY(0, 0);
        }
        private void btn1600_1600_Click(object sender, EventArgs e)
        {

            FahrtXY(1600, 1600);
        }

        private void btn6000_1000_Click(object sender, EventArgs e)
        {
            FahrtXY(6000, 1000);
        }

        private void btn3200_1600_Click(object sender, EventArgs e)
        {
            FahrtXY(3200, 1600);
        }

        private void btn2000_1000_Click(object sender, EventArgs e)
        {
            FahrtXY(2000, 1000);
        }

        private void btn4000_800_Click(object sender, EventArgs e)//geändert in 2000/4000
        {
            FahrtXY(2000, 4000);
        }

        private void btn9900_700_Click(object sender, EventArgs e)
        {
            FahrtXY(9900, 700);
        }

        private void btnTextboxSend_Click(object sender, EventArgs e)
        {
            FahrtXY(5000, 3000);
        }
        private void btnLabelSend_Click(object sender, EventArgs e)
        {
            FahrtXY(2000, 5500);
        }
        private void Form1_Load(object sender, EventArgs e)
        {
            
        }
        //-----------------------------------------------------------
        private void btnTextBoxTest_Click(object sender, EventArgs e)
        {
            tbReceived.Clear();
            tbxTest1.Clear();
            tbxTest2.Clear();
            tbxTest3.Clear();
            tbxTest4.Clear();
            serialPort1.Write("14,0,0,0,0");
            Sleep(1500);
            var Zeile = this.tbReceived.Lines;
            try
            {
                this.tbxTest1.Text = Zeile[0];
                this.tbxTest2.Text = Zeile[1];
                this.tbxTest3.Text = Zeile[2];
                this.tbxTest4.Text = Zeile[3];
            }
            catch //(Exception ex)
            {
                MsgBox("Es sind noch keine Werte vorhanden");
            }
        }
        //-----------------------------------------------------------
        private void btnTestVar_Click(object sender, EventArgs e)
        {
            FahrtXY(4650, 7280);
        }
        //-----------------------------------------------------------
        private void txbRunden1_MouseDoubleClick(object sender, MouseEventArgs e)
        {
            serialPort1.Write("15," + txbRunden1.Text + ",0,0,0");
        }
        //-----------------------------------------------------------
        private void btnStaticTime_Click(object sender, EventArgs e)
        {
            int Pausenzeit = 5000;
            tbxTest3.Text = "Pause: "+ Pausenzeit.ToString();
            serialPort1.Write("1,600,700,0,0");
            Sleep(Pausenzeit);
            serialPort1.Write("1,0,0,0,0");
            Sleep(Pausenzeit);
            serialPort1.Write("1,400,350,0,0");
            Sleep(Pausenzeit);
            serialPort1.Write("1,0,0,0,0");
        }
        //-----------------------------------------------------------
        private void txbRunden1_KeyPress(object sender, KeyPressEventArgs e)
        {
            if (e.KeyChar == 13)
            {
                serialPort1.Write("15," + txbRunden1.Text + ",0,0,0");
                e.Handled = true;//DingDong ausschalten...
            }
        }
        public static int map(int value, int fromLow, int fromHigh, int toLow, int toHigh)
        {
            return (value - fromLow) * (toHigh - toLow) / (fromHigh - fromLow) + toLow;
        }//END private static int map()------------------------------

        private void btnReadCurrentAngles_Click(object sender, EventArgs e)
        {
            tbReceived.Clear();                      //ServoReadAngles()   case 16:
            serialPort1.Write("16,0,0,0,0");
            Sleep(1500);
            var Zeile = this.tbReceived.Lines;
            try
            {
                this.txbServoActualAngleMin.Text = Zeile[0];
                this.txbServoActualAngleMax.Text = Zeile[1];
            }
            catch //(Exception ex)
            {
                MsgBox("Es sind noch keine Werte vorhanden");
            }
            //Sleep(1000);
            tbReceived.Clear();
        }

        private void btnWriteDefaultAngles_Click(object sender, EventArgs e)
        {
            serialPort1.Write("17,10,170,0,0");    //ServoUpdateAngles()   case 17:
            txbServoUpdatelAngleMin.Text = "10";
            txbServoUpdatelAngleMax.Text= "170";            
        }

        private void btnWriteUpdateAngles_Click(object sender, EventArgs e)
        {
            this.serialPort1.Write("17," + txbServoUpdatelAngleMin.Text + "," + txbServoUpdatelAngleMax.Text + ",0,0");
        }

        //-----------------------------------------------------------
        private void btnWhileSchleife_Click(object sender, EventArgs e)
        {
            FahrtXY(1200, 1300);
            Sleep(sleepPause);
            Sleep(1000); //zusätzliche Hilfe, map-Funktion funktioniert nicht immer in FahrtXY()

            FahrtXY(0, 0);
            Sleep(sleepPause);
            Sleep(1000); //zusätzliche Hilfe, map-Funktion funktioniert nicht immer in FahrtXY()

            FahrtXY(1500, 600);
            Sleep(sleepPause);
            Sleep(1000); //zusätzliche Hilfe, map-Funktion funktioniert nicht immer in FahrtXY()

            FahrtXY(0, 0);
            Sleep(sleepPause);
            Sleep(1000); //zusätzliche Hilfe, map-Funktion funktioniert nicht immer in FahrtXY()

            FahrtXY(2300, 1700);
            Sleep(sleepPause);
            Sleep(2000); //zusätzliche Hilfe, map-Funktion funktioniert nicht immer in FahrtXY()

            FahrtXY(0, 0);
            Sleep(sleepPause);
        }
        //-----------------------------------------------------------
        public void FahrtXY(int Steps_X, int Steps_Y)
        {
            //die map-Funktion ist noch nicht ausgereift. Mehrere Motorläufe hintereinander
            //funktionieren nicht immer...

            sollPosition_X = Steps_X;
            sollPosition_Y = Steps_Y;
            stepsToGo_X = Math.Abs(sollPosition_X - istPosition_X);
            stepsToGo_Y = Math.Abs(sollPosition_Y - istPosition_Y);
            SummeSteps_XY = stepsToGo_X + stepsToGo_Y;
            tbxSumSteps.Text = SummeSteps_XY.ToString();
            //#######################################################
            //this.serialPort1.Write("1,Steps_X, Steps_Y,0,0"); //--> So funktioniert es nicht...
            this.serialPort1.Write("1," + Steps_X + "," + Steps_Y + ",0,0");//zur seriellen Schnittstelle
            //#######################################################
            istPosition_X = sollPosition_X;
            istPosition_Y = sollPosition_Y;

            //map-Funktion ##########################################
            sleepPause = map(SummeSteps_XY, 500, 10000, 1000, 4500);
            //#######################################################

            //tbxSleepPause.Text = sleepPause.ToString();
            tbxTest4.Text = "Pause: " + sleepPause.ToString();             
        }
        //-----------------------------------------------------------

    }//END public partial class Form1 : Form
}//END namespace AdeTeK_Plottertest
